File: app/src/main/java/ru/application/homemedkit/fragments/FragmentSettings.java
package ru.application.homemedkit.fragments;

import static android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import static ru.application.homemedkit.helpers.ConstantsHelper.CHECK_EXP_DATE;
import static ru.application.homemedkit.helpers.ConstantsHelper.SETTINGS_CHANGED;
import static ru.application.homemedkit.helpers.SettingsHelper.APP_THEME;
import static ru.application.homemedkit.helpers.SettingsHelper.LANGUAGE;
import static ru.application.homemedkit.helpers.SettingsHelper.LIGHT_PERIOD;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.preference.PreferenceFragmentCompat;

import ru.application.homemedkit.R;
import ru.application.homemedkit.helpers.SettingsHelper;

public class FragmentSettings extends PreferenceFragmentCompat implements OnSharedPreferenceChangeListener {

    private SettingsHelper preferences;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        preferences = new SettingsHelper(requireActivity().getBaseContext());
    }

    @Override
    public void onCreatePreferences(@Nullable Bundle savedInstanceState, @Nullable String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, @Nullable String key) {
        if (key != null) {
            var value = sharedPreferences.getAll().get(key);

            switch (key) {
                case LANGUAGE -> {
                    requireActivity().getIntent().putExtra(SETTINGS_CHANGED, true);
                    preferences.changeLanguage((String) value);
                }
                case APP_THEME -> {
                    requireActivity().getIntent().putExtra(SETTINGS_CHANGED, true);
                    preferences.setAppTheme((String) value);
                }
                case CHECK_EXP_DATE ->
                        preferences.setExpDateChecker(requireActivity(), (Boolean) value);
                case LIGHT_PERIOD -> preferences.setLightPeriod((Boolean) value);
            }
        }
    }
}

File: app/src/main/java/ru/application/homemedkit/fragments/FragmentSettings.kt
package ru.application.homemedkit.fragments

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_NO
import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_YES
import androidx.appcompat.app.AppCompatDelegate.setApplicationLocales
import androidx.appcompat.app.AppCompatDelegate.setDefaultNightMode
import androidx.core.os.LocaleListCompat.forLanguageTags
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import ru.application.homemedkit.R
import ru.application.homemedkit.activities.MainActivity
import ru.application.homemedkit.activities.MainActivity.Companion.preferences
import ru.application.homemedkit.helpers.CHECK_EXP_DATE
import ru.application.homemedkit.helpers.KEY_APP_THEME
import ru.application.homemedkit.helpers.KEY_DOWNLOAD
import ru.application.homemedkit.helpers.KEY_FRAGMENT
import ru.application.homemedkit.helpers.KEY_LANGUAGE
import ru.application.homemedkit.helpers.KEY_LIGHT_PERIOD
import ru.application.homemedkit.helpers.KEY_ORDER
import ru.application.homemedkit.helpers.PAGES
import ru.application.homemedkit.helpers.SETTINGS_CHANGED
import ru.application.homemedkit.helpers.SORTING
import ru.application.homemedkit.helpers.THEMES

class FragmentSettings : PreferenceFragmentCompat(), OnSharedPreferenceChangeListener {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        sharedPreferences?.let { preferences ->
            key?.let { string -> val value = preferences.all[string]
                when (string) {
                    KEY_LIGHT_PERIOD -> {
                        preferences.edit().putBoolean(KEY_LIGHT_PERIOD, value as Boolean).apply()
                    }

                    KEY_ORDER -> {
                        preferences.edit().putString(KEY_ORDER, string).apply()
                    }

                    KEY_DOWNLOAD -> {
                        preferences.edit().putBoolean(KEY_DOWNLOAD, value as Boolean).apply()
                    }

                    CHECK_EXP_DATE -> {
                        preferences.edit().putBoolean(CHECK_EXP_DATE, value as Boolean).apply()
                        (requireActivity() as MainActivity).setExpirationChecker()
                    }

                    KEY_LANGUAGE -> {
                        preferences.edit().putString(KEY_LANGUAGE, value.toString()).apply()
                        activity?.intent?.putExtra(SETTINGS_CHANGED, true)
                        setApplicationLocales(forLanguageTags(value.toString()))
                    }

                    KEY_APP_THEME -> {
                        preferences.edit().putString(KEY_APP_THEME, value.toString()).apply()
                        activity?.intent?.putExtra(SETTINGS_CHANGED, true)

                        when (value) {
                            THEMES[1] -> setDefaultNightMode(MODE_NIGHT_NO)
                            THEMES[2] -> setDefaultNightMode(MODE_NIGHT_YES)
                            else -> setDefaultNightMode(MODE_NIGHT_FOLLOW_SYSTEM)
                        }
                    }

                    else -> {}
                }
            }
        }
    }

    fun getHomePage() = preferences.getString(KEY_FRAGMENT, PAGES[0])!!
    fun getSortingOrder() = preferences.getString(KEY_ORDER, SORTING[0])!!
    fun getDownloadNeeded() = preferences.getBoolean(KEY_DOWNLOAD, false)
    fun getLightPeriod() = preferences.getBoolean(KEY_LIGHT_PERIOD, true)
    fun getAppTheme() = when (preferences.getString(KEY_APP_THEME, THEMES[2])) {
        THEMES[1] -> setDefaultNightMode(MODE_NIGHT_NO)
        THEMES[2] -> setDefaultNightMode(MODE_NIGHT_YES)
        else -> setDefaultNightMode(MODE_NIGHT_FOLLOW_SYSTEM)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)

        preferences = PreferenceManager.getDefaultSharedPreferences(context)
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences?.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences?.registerOnSharedPreferenceChangeListener(this)
    }
}